src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoverFromLocalAssemblyContigAlignmentsSpark.java

```diff
@@ -146,20 +151,60 @@ protected void runTool(final JavaSparkContext ctx) {
                         cnvCallsBroadcast,
                         getReads(), getHeaderForReads(), getReference(), localLogger);
 
-        final EnumMap<RawTypes, JavaRDD<AssemblyContigWithFineTunedAlignments>> contigsByPossibleRawTypes =
+        final AssemblyContigsClassifiedByAlignmentSignatures contigsByPossibleRawTypes =
                 preprocess(svDiscoveryInputData, writeSAMFiles);
 
         dispatchJobs(contigsByPossibleRawTypes, svDiscoveryInputData);
     }
 
     //==================================================================================================================
 
+    public static final class AssemblyContigsClassifiedByAlignmentSignatures {
+        final JavaRDD<AssemblyContigWithFineTunedAlignments> ambiguous;
+        final JavaRDD<AssemblyContigWithFineTunedAlignments> incomplete;
+        final JavaRDD<AssemblyContigWithFineTunedAlignments> misassembly;
+        final JavaRDD<AssemblyContigWithFineTunedAlignments> simple;
+        final JavaRDD<AssemblyContigWithFineTunedAlignments> complex;
+
+        private AssemblyContigsClassifiedByAlignmentSignatures(final JavaRDD<AssemblyContigWithFineTunedAlignments> ambiguous,
+                                                               final JavaRDD<AssemblyContigWithFineTunedAlignments> incomplete,
+                                                               final JavaRDD<AssemblyContigWithFineTunedAlignments> misassembly,
+                                                               final JavaRDD<AssemblyContigWithFineTunedAlignments> simple,
+                                                               final JavaRDD<AssemblyContigWithFineTunedAlignments> complex) {
+            this.ambiguous = ambiguous;
+            this.incomplete = incomplete;
+            this.misassembly = misassembly;
+            this.simple = simple;
+            this.complex = complex;
+        }
+
+        /**
+         * Write SAM file, if requested, for original alignments of contigs recognized as "Ambiguous", "Incomplete", and "MisAssemblySuspect"
+         * TODO: 11/17/17 salvation on assembly contigs that 1) has ambiguous "best" configuration, and 2) has incomplete picture; and flag accordingly
+         *
+         * TODO: 3/4/18 a bug is present here that even though only one alignment has not-bad MQ, it could contain a large gap,
+         *      depending on the behavior of the other gap-less bad mappings,
+         *      we may end up classifying the whole contig as incomplete, or signal-less,
+         *      we should keep the single not-bad mapping and annotate accordingly
+         */
+        private void writeSAMfilesForSuspicious(final String outputPrefix, final JavaRDD<GATKRead> assemblyRawAlignments,
```

Comment:
This seems like a very inefficient way to write these three files. You end up calling `collect` on the RDD three different times and then traversing the local collection three times. Why not make a map of contig name to bam file, collect the rdd once, and then traverse the local collection once, writing each read to the appropriate bam file from the map?